### ISSUE TITLE
Multiple OSMD instances inevitably share their settings

### ISSUE THREAD

**Issue by dnlfrst** [opensheetmusicdisplay/opensheetmusicdisplay#559]
# Issue
***Note:*** *this issue is illustrated in a [minimal working example](https://codesandbox.io/s/osmd-issue-558-t4503).*

## Assumption
- I want to display a full-blown music sheet (e.g. _Beethoven's_ "_[An die ferne Geliebte](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/blob/develop/test/data/Beethoven_AnDieFerneGeliebte.xml)_") and a fragment of the very same (e.g. the very first measure) on the same page.
- As the fragment consists of a single measure only, I want to have it displayed compactly and not show the title, the composer, or the part names.

## Implementation Detail
Both, the full-blown music sheet and the fragment, receive a separate [`OpenSheetMusicDisplay`](https://opensheetmusicdisplay.github.io/classdoc/classes/opensheetmusicdisplay.html) instance with their own container.  
Now, since the fragment should be compact and not receive a title, a composer, or part name, I remove them using the appropriate options for the respective `OpenSheetMusicDisplay` instance. The full-blown music sheet does not receive any options (i.e. the default options):

```javascript
const musicSheetFragmentOptions = {
  drawComposer: false,
  drawPartAbbreviations: false,
  drawPartNames: false,
  drawSubtitle: false,
  drawTitle: false,
  darwingParameters: DrawingParametersEnum.compact
};

const musicSheetFragment = new OpenSheetMusicDisplay(
  "music-sheet-fragment",
  musicSheetFragmentOptions
);
const fullMusicSheet = new OpenSheetMusicDisplay("full-music-sheet");
```

The [result of the above](https://codesandbox.io/s/osmd-issue-558-t4503) is that both `OpenSheetMusicDisplay` instances receive the default options while the given ones are ignored.

# Failed Attempt at a Solution
- Setting the default options on the second instance explicitly,
- lazily setting the options after rendering, or
- changing the rendering order

does not do the trick. All of which does not come to my surprise since the [`engraving rules`](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/blob/develop/src/MusicalScore/Graphical/EngravingRules.ts) as the underlying driver of the settings are implemented as a [Singleton](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/blob/develop/src/MusicalScore/Graphical/EngravingRules.ts#L437).

# Questions
- Why do the settings (cf. [`EngravingRules`](https://github.com/opensheetmusicdisplay/opensheetmusicdisplay/blob/develop/src/MusicalScore/Graphical/EngravingRules.ts)) behave like a Singleton? What are the benefits in doing so?
- Is there any way around the tangled settings such that each OSMD instance can receive their own options?

**Comment by sschmidTU** [comment]
First of all, great issue report!

The EngravingRules behave like a Singleton because we need to call them from many parts of the code, and it is convenient to use the Singleton pattern for that. But of course in the Browser two instances of OSMD will share the Singleton.
I had not thought of the use case of rendering two instances of OSMD in the same document before, though of course it completely makes sense. 

This is something we'll have to fix. My first impulse is to link the EngravingRules object to the OSMD object, so an EngravingRules object is created for every OSMD instance.

There may not currently be a workaround. Though have you tried setting the options each time before rendering each OSMD object? something like this:
```js
osmd1.setOptions(renderingOptions1);
osmd1.render()
osmd2.setOptions(renderingOptions2); // will also set options for next render of osmd1
osmd2.render();
```

Though since this is an important and not very hard issue to fix, we may go for a fix for this Singleton issue very soon (hopefully on the weekend, releasing the next OSMD version as well).

**Comment by dnlfrst** [comment]
Thanks for the insight!

---

> There may not currently be a workaround. Have you tried setting the options each time before rendering each OSMD object? something like this:

> ```javascript
> osmd1.setOptions(renderingOptions1);
> osmd1.render()
> osmd2.setOptions(renderingOptions2); // will also set options for next render of osmd1
> osmd2.render();
> ```

I thought I tried this as well, turned out I didn't! Setting the options for both instance before their respective `render` call indeed allows to set the options "independently".  
**However**, this still requires to set every single option explicitly. That is, the default options would need to be specified and cannot be set implicitly (i.e. left out). See [the modified example](https://codesandbox.io/s/osmd-issue-558-t4503) for illustration.  
Therefore, if the problem can be fixed without too much hassle that would still be the best solution. At least, there's a workaround available until the next release.

**Comment by AlbertHart** [comment]
I did find a simple way to render the same XML - with different option in this issue: #594 

By rendering to the first container, saving its innerHTML (which can changed even when rendering again to the second container) and the restoring the first container when the second one is done,

As I am typing this I see that sschmidtTU has added this issue to my previous issue. I'll go read that now and see if there is a solution.

But I, also, would like an easier solution.

**Comment by sschmidTU** [comment]
One way of identifying an EngravingRules instance with an OSMD instance is to save the EngravingRules reference in the MusicSheet (and not make EngravingRules static anymore).

The MusicSheet is already referenced in many classes, and we could add a reference in the SourceMeasure for example, which you can get to from most rendering code.
(suggestion by @matt-uib)

**Comment by sschmidTU** [comment]
@AlbertHart FYI, this is now fixed and released in OSMD 0.7.5.
